src/zarr/storage/_obstore.py

```diff
@@ -0,0 +1,419 @@
+            )
+        buf = value.to_bytes()
+        await obs.put_async(self.store, key, buf)
+
+    async def set_if_not_exists(self, key: str, value: Buffer) -> None:
+        # docstring inherited
+        import obstore as obs
+
+        self._check_writable()
+        buf = value.to_bytes()
+        with contextlib.suppress(obs.exceptions.AlreadyExistsError):
+            await obs.put_async(self.store, key, buf, mode="create")
+
+    @property
+    def supports_deletes(self) -> bool:
+        # docstring inherited
+        return True
+
+    async def delete(self, key: str) -> None:
+        # docstring inherited
+        import obstore as obs
+
+        self._check_writable()
+        await obs.delete_async(self.store, key)
+
+    @property
+    def supports_partial_writes(self) -> bool:
+        # docstring inherited
+        return False
+
+    async def set_partial_values(
+        self, key_start_values: Iterable[tuple[str, int, BytesLike]]
+    ) -> None:
+        # docstring inherited
+        raise NotImplementedError
+
+    @property
+    def supports_listing(self) -> bool:
+        # docstring inherited
+        return True
+
+    def list(self) -> AsyncGenerator[str, None]:
+        # docstring inherited
+        import obstore as obs
+
+        objects: ListStream[list[ObjectMeta]] = obs.list(self.store)
+        return _transform_list(objects)
+
+    def list_prefix(self, prefix: str) -> AsyncGenerator[str, None]:
+        # docstring inherited
+        import obstore as obs
+
+        objects: ListStream[list[ObjectMeta]] = obs.list(self.store, prefix=prefix)
+        return _transform_list(objects)
+
+    def list_dir(self, prefix: str) -> AsyncGenerator[str, None]:
+        # docstring inherited
+        import obstore as obs
+
+        coroutine = obs.list_with_delimiter_async(self.store, prefix=prefix)
+        return _transform_list_dir(coroutine, prefix)
+
+
+async def _transform_list(
+    list_stream: ListStream[list[ObjectMeta]],
+) -> AsyncGenerator[str, None]:
+    """
+    Transform the result of list into an async generator of paths.
+    """
+    async for batch in list_stream:
+        for item in batch:
+            yield item["path"]
+
+
+async def _transform_list_dir(
+    list_result_coroutine: Coroutine[Any, Any, ListResult[list[ObjectMeta]]], prefix: str
+) -> AsyncGenerator[str, None]:
+    """
+    Transform the result of list_with_delimiter into an async generator of paths.
+    """
+    list_result = await list_result_coroutine
+
+    # We assume that the underlying object-store implementation correctly handles the
+    # prefix, so we don't double-check that the returned results actually start with the
+    # given prefix.
+    prefixes = [obj.lstrip(prefix).lstrip("/") for obj in list_result["common_prefixes"]]
+    objects = [obj["path"].lstrip(prefix).lstrip("/") for obj in list_result["objects"]]
```

Comment:
Would `.removeprefix` be correct here? (https://peps.python.org/pep-0616/)

Reply:
nice catch~ `lstrip(prefix)` should be wrong . Worrying that a Store test hasn't caught that.

Reply:
Closed in [`0735a43` (#1661)](https://github.com/zarr-developers/zarr-python/pull/1661/commits/0735a437bf38740b2fe47a613652eb2bb17e486c)